### PR TITLE
Hide keras model so that clone function from sklearn works

### PR DESCRIPTION
**Pull request by jnlaia**
There are actually several changes I've made. I believe they can be useful for other people as well.

#### Hide keras model within the wrapper so that sklearn clone function works
It is essential that the sklearn.base.clone function works so that keras can be used with sklearn functions, such as BaggingClassifier or GridSearchCV, etc. Currently, the clone function has issues with keras. The way that function works is that it calls the get_params method from the classifier, and then does it recursively to each of the parameters in case that method is present. If not, it just deep copies the parameter. 
It turns out that sklearn should only see the get_params from the wrapper class, but not the get_params from the keral model itself, as the latter doesn't follow the sklearn API (and it is not supposed to). By hiding this model inside a Hideout class, we can fix this issue.

#### Add a class_weight = 'balanced' (as in sklearn)
This way, when calling fit, the class_weight will be computed from the class frequency.

#### Created a decorator do use in fit(X, y) and score(X, y)
This code was already in the fit function. However it should also be in the score function. To avoid repeating it, I created a decorator to do the job.
Recall, this code turns a vector y into a one hot encoded output in case the loss_function is 'categorical_crossentropy'

#### Some minor fixes
There were some other minor changes, like moving some Classifier specific code from the BaseWrapper into the KerasClassifier, and some other minor stuff.

I've left the KerasRegressor code intact.

### CODE REVIEW

**Comment by jnlaia** [comment]
I've added some tests in a different branch: https://github.com/jnlaia/keras/blob/sklearn_tests/tests/test_sklearn.py

Some of those tests require sklearn to run. Let me know if there is interest in adding them to the main repo.

**Comment by ekerazha** [comment]
@jnlaia 
It seems like it doesn't correctly work with sklearn.calibration.CalibratedClassifierCV()
Any insight?

File: keras/wrappers/scikit_learn.py [hunk 1]
```diff
@@ -6,6 +6,18 @@
 from ..utils.np_utils import to_categorical
 
 
+class Hideout(object):
```
Comment (fchollet) [inline keras/wrappers/scikit_learn.py]: Since this is only for use by one method in one class, you should define this class in the scope of that method.

Additionally, it doesn't seem (intuitively) to be a very clean pattern. Are you sure this is the best way to proceed?

File: keras/wrappers/scikit_learn.py [hunk 2]
```diff
@@ -67,15 +91,26 @@ def get_params(self, deep=True):
         Parameters
         ----------
```
Comment (fchollet) [inline keras/wrappers/scikit_learn.py]: It's outside of the scope of the PR, but I am reminded that this part of the code should be converted to the new style of dosctrings. This would allow us to autogenerate the documentation.

File: keras/wrappers/scikit_learn.py [hunk 3]
```diff
@@ -149,8 +179,108 @@ class KerasClassifier(BaseWrapper):
...
+                                              show_accuracy,
+                                              validation_split, 
+                                              validation_data, 
+                                              callbacks,
+                                              verbose)
+
+        self.class_weight = class_weight
+
+
+    def __ohe_output(function):
```
Comment (fchollet) [inline keras/wrappers/scikit_learn.py]: The name is obscure.

**Comment by fchollet** [comment]
You can move here any test that doesn't require sklearn.

File: keras/wrappers/scikit_learn.py (hunk 1 above)
Comment (jnlaia) [inline keras/wrappers/scikit_learn.py]: I'm not sure it is the best way to proceed. It is definitely the best one I have though about. The most straightforward solution, which I didn't want to pursue, would be to change the name of the get_params method all over keras core code. At least in the current solution all the changes are confined to the scikit_learn.py file.

I would be happy to hear different approaches, if there are any.

File: keras/wrappers/scikit_learn.py (hunk 3 above)
Comment (jnlaia) [inline keras/wrappers/scikit_learn.py]: The name ohe_output comes from "One Hot Encode the output". I agree it is a little obscure. Maybe "output_to_categorical" or "categorize_output"? Do you have any suggestion?

Comment (fchollet) [inline keras/wrappers/scikit_learn.py]: I would suggest dropping the decorator and instead repeating the code twice. You could abstract it in a helper function if that's really needed. I don't think a decorator is helpful in terms of readability in this case.

File: keras/wrappers/scikit_learn.py [hunk 4]
```diff
@@ -115,23 +151,17 @@ def fit(self, X, y):
...
         """
-        if len(y.shape) == 1:
-            self.classes_ = list(np.unique(y))
-            if self.loss == 'categorical_crossentropy':
-                y = to_categorical(y)
-        else:
-            self.classes_ = np.arange(0, y.shape[1])
 
-        self.compiled_model_ = copy.deepcopy(self.model)
+        self.compiled_model_ = copy.deepcopy(self.model.model)
```
Comment (fchollet) [inline keras/wrappers/scikit_learn.py]: In general private attributes should prefixed with `_` rather than the opposite.

File: keras/wrappers/scikit_learn.py (hunk 1 above)
Comment (fchollet) [inline keras/wrappers/scikit_learn.py]: Could you achieve the same with a different class attribute under a different name? Is the nesting essential?

Comment (jnlaia) [inline keras/wrappers/scikit_learn.py]: The name is not important. The fact that the class attribute "model" has a get_params method is the problem here. 

Notice that "compiled_model_" doesn't have that problem because that attribute is not returned by the BaseWrapper get_params method, and therefore scikit learn will not look at it.

File: keras/wrappers/scikit_learn.py (hunk 3 above)
Comment (jnlaia) [inline keras/wrappers/scikit_learn.py]: I disagree with you on this one. Even if this code was showing only once, I would still be happy yo put it in a decorator. This code has nothing to do with the logic of the fit or score methods. It rather dresses the input of those functions. From my point of view this decorator removes unnecessary code from within the fit and score functions, and therefore makes them more readable. It separates the code into distinct logical parts.

**Comment by jnlaia** [comment]
In this PR I've avoided making any decision regarding the scikit learn API. Instead, I've build upon the one that was already in keras, and made some small changes to fix it. As it is now, keras scikit learn API is broken.

I still think it would be helpful to have, at some point, a deeper discussion on this API. For example, wouldn't it be nice to have some of the parameters from the class attribute "model" exposed to the outside, so that we would be able to do grid searches on them as well? As it is, it is not immediate to do a grid search where one changed the dropout or the geometry of the neural net. Changing the current API is outside of the scope of this PR though.

**Comment by jnlaia** [comment]
I will only be able to implement the changes you suggested by the end of the week that is now starting, or the following one.

File: keras/wrappers/scikit_learn.py (hunk 4 above)
Comment (jnlaia) [inline keras/wrappers/scikit_learn.py]: I'll change this. The reason why the underscore was at the end is related to a sklearn convention: every public attribute that is set in the fit method has an undescore in the end. I guess one can argue that the compiled_model should be a private attribute instead, so I'll change the position of the underscore.

File: keras/wrappers/scikit_learn.py (hunk 2 above)
Comment (jnlaia) [inline keras/wrappers/scikit_learn.py]: I'll leave this for afterwards, if you don't mind.
Also, Is there a place where I can see the new convention, other that the code itself?

File: keras/wrappers/scikit_learn.py (hunk 3 above)
Comment (jnlaia) [inline keras/wrappers/scikit_learn.py]: I haven't changed this part as I have some hope I've changed your mind. Let me know if not, and I'll change it accordingly.

File: tests/test_sklearn.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
+import keras
+from keras.layers.core import Dense, Dropout, Activation
+from keras.models import Sequential
+from keras.optimizers import SGD
```
Comment (ekerazha) [inline tests/test_sklearn.py]: SGD is neved used.

**Comment by fchollet** [comment]
Closing outdated PRs following the release of Keras 1.0.
